internals/src/hex/display.rs

```diff
@@ -36,6 +36,17 @@ pub trait DisplayHex: Copy + sealed::IsRef {
     /// Avoids the requirement to import the `Case` type.
     fn display_upper_hex(self) -> Self::Display { self.display_hex(Case::Upper) }
 
+    /// Create a hex-encoded string.
+    ///
+    /// This may be faster than `.display_hex().to_string()` because it uses `reserve_suggestion`.
+    #[cfg(feature = "alloc")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "alloc")))]
+    fn to_hex_string(self, case: Case) -> String {
+        let mut string = String::new();
+        self.append_hex_to_string(case, &mut string);
+        string
+    }
+
```

Comment:
First of all, the ordering should be "most useful first". `display_*` is considered more useful because it avoids allocations and `_{upper|lower}_` is more useful because it avoids annoying imports.

But I realized this API is more complicated than needed. It should be `Self::Display: LowerHex + UpperHex` so that people formatting it can simply write `"{:x}"`. `to_*_string` are still needed because it is useful where you'd use `.to_string()`. Given this new information I'd rename the trait:


```rust
pub trait AsHex: Copy + sealed::IsRef { // or should we really say `SliceExt`? I doubt we will ever impl it for a type that doesn't deref into `&[u8]`
    type Hex: fmt::LowerHex + fmt::UpperHex;
    fn as_hex(self) -> Self::Hex;

    // remaining string and reserve_suggestion methods unchanged
}
```

There's still an edge case like `serde` `collect_str()` (which we should change to `serialize_str` for fixed-sized types) which can be solved simply by `format_args!("{:x}", value)` or ,in perf-sensitive applications, by having a wrapper struct forwarding to `LowerHex`. If this becomes frequent we may add a helper here.